azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -231,9 +229,30 @@ static ExecutableFlow getFlowToRestart(final ExecutableFlow flow,
       restartedStatuses.add(Status.EXECUTION_STOPPED.name());
     }
 
+    // flow can only retry if custom-retry-times not reach the limit
+    // and for the case where the retry-times not defined, default to 1
+    int flowMaxRetryLimit = 0;
+    if (flowParams.containsKey(FlowParameters.FLOW_PARAM_MAX_RETRIES)){
+      flowMaxRetryLimit = Integer.parseInt(flowParams.get(FlowParameters.FLOW_PARAM_MAX_RETRIES));
+    }
+    else if (!restartedStatuses.isEmpty()){
```

Comment:
Would line 221 return a set with one element which is an empty String when the user didn't configure FLOW_PARAM_ALLOW_RESTART_ON_STATUS?

Reply:
good point, let me add filter